datadog/dogstatsd/base.py

```diff
@@ -453,16 +450,16 @@ def open_buffer(self, max_buffer_size=None):
 
         self._manual_buffer_lock.acquire()
 
+        # XXX Remove if `disable_buffering` default is changed to False
+        self._send = self._send_to_buffer
```

Comment:
This stuff is not thread-safe. I presume it never really was and I assume that's beyond the expected use-cases.

Reply:
@truthbk Generally things should be thread-safe due to various locks on the socket/buffer/context ops but you may get unexpectedly-buffered metrics if some threads are opening/closing buffers while another thread is just writing metrics without a context manager or `open_buffer`. Overall though, there should be no data loss or exceptions. As for the setter here, the lock + GIL makes this effectively an atomic operation. We do have a test around at least part of this [here](https://github.com/DataDog/datadogpy/blob/33f727cc51d637860a2a46677bf7151c8feb10fa/tests/unit/dogstatsd/test_statsd.py#L711). If I'm missing something though, I'll fix up the PR for sure.

> I presume it never really was and I assume that's beyond the expected use-cases.

Somewhat. The way it was coded originally was not really thread safe but the changes made subsequently over this year have closed up most of the glaring exception issues. The buffering-by-default fixed the `self._send =` GIL reliance but we're backing that part out with this PR for the time being 😢 .